### ISSUE TITLE
Loop unrolling and pairwise summation

### ISSUE THREAD

**Issue by kwgoodman** [kwgoodman/bottleneck#117]
I'm playing around with loop unrolling in the unroll branch. I did some unrolling of bn.nansum. Here's the performance on my computer (nansum2 is the unrolled one):

With float64:
```
In [1]: bn.bench(functions=['nansum', 'nansum2'])
Bottleneck performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is NumPy time divided by Bottleneck time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN     no NaN      NaN        NaN    
                   (10,)   (1000,1000)   (10,)   (1000,1000)
    nansum         32.1        2.6       32.3        7.4
    nansum2        41.9        3.6       42.1        9.2
```

With int64:
```
In [3]: bn.bench(functions=['nansum', 'nansum2'], dtype='int64')
Bottleneck performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is NumPy time divided by Bottleneck time
    NaN means approx one-third NaNs; int64 and axis=-1 are used

                 no NaN     no NaN      NaN        NaN    
                   (10,)   (1000,1000)   (10,)   (1000,1000)
    nansum         13.1        1.1       12.7        1.1
    nansum2        19.5        1.4       19.8        1.4
```

And here are some more timings:

```
In [4]: a = np.random.rand(100000000)
In [5]: timeit bn.nansum(a)
10 loops, best of 3: 78.5 ms per loop
In [6]: timeit bn.nansum2(a)
10 loops, best of 3: 59.2 ms per loop
```

Neat.

**Comment by jennolsen84** [comment]
Perhaps even try to build the code with `-O3` and `-funroll-loops` (if you're using GCC locally).  You can also try `-f-unroll-all-loops`.  The `-O3` will also unlock `--ftree-vectorize`, and then you can examine the output using `-ftree-vectorizer-verbose=2` to see if the autovectorization is happening (nansum is a perfect place where autovectorization would help).

For clang, I think `-O1` will unroll the loops.

You can also try profile guided optimization.  This has been known to provide substantial benefit.

**Comment by kwgoodman** [comment]
Good idea. Here are the timings with the default gcc flags:
```
In [1]: bn.bench(functions=['nansum', 'nansum2'])
Bottleneck performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is NumPy time divided by Bottleneck time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN     no NaN      NaN        NaN    
                   (10,)   (1000,1000)   (10,)   (1000,1000)
    nansum         29.5        2.6       30.6        7.4
    nansum2        39.3        3.6       41.9        9.2
```

with -O3 -funroll-loops:
```
    nansum         34.9        2.5       31.6        7.7
    nansum2        35.2        3.6       35.1        9.7
```

with -O3 -funroll-all-loops:
```
    nansum         34.5        2.5       36.2        7.8
    nansum2        39.7        3.6       39.6        9.8

```

Doesn't seem like the flags do much of anything on this code/computer.

BTW, in addition to loop unrolling I am also doing pairwise [summation](https://en.wikipedia.org/wiki/Pairwise_summation). So in addition to speed we are getting better sums some of the time. Pairwise summation should also help prevent nanmean from overflowing during the summation stage.

If I unroll the loops without pairwise summation the speed advantage goes away!
```
    nansum         36.8        2.5       36.9        7.3
    nansum2        40.4        2.5       40.7        7.9
```

So, yeah, I don't understand the where the speed up come from.

**Comment by kwgoodman** [comment]
Looking a little closer at the definition of pairwise summation I see that that is not what I am doing. The number of sub-sums I calculate (4) does not depend on the size of the array (as long as there are at least 8 elements along the reduction axis). So I have not done the divide and conquer part of pairwise summation. I guess we could call it pairwise summation with a limit of the max np.intp value before switching to divide and conquer.

**Comment by kwgoodman** [comment]
I need help running the bottleneck benchmark on the unroll branch. Here's what I get on 64-bit ubuntu 14.04:
```
In [1]: bn.bench2()
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         1.41       1.25
```

So those results mean that bottleneck master nansum runtime divided by unrolled nansum is 1.41 with no NaNs. I'm wondering if those results are just luck with my compiler/cache/etc. So I'd like to see the performance on your computer.

Some background:

Here's part of the loop unroll for nansum:
```        
    elif length >= 8:
            while PyArray_ITER_NOTDONE(ita):
                x[0] = 0
                x[1] = 0
                x[2] = 0
                x[3] = 0
                p = <char*>pid(ita)
                ai = (<DTYPE0_t*>p)[0]
                if ai == ai: x[0] += ai
                ai = (<DTYPE0_t*>(p + 1 * stride))[0]
                if ai == ai: x[1] += ai
                ai = (<DTYPE0_t*>(p + 2 * stride))[0]
                if ai == ai: x[2] += ai
                ai = (<DTYPE0_t*>(p + 3 * stride))[0]
                if ai == ai: x[3] += ai
                for i in range(4, length - (length % 4), 4):
                    ai = (<DTYPE0_t*>(p + i * stride))[0]
                    if ai == ai: x[0] += ai
                    ai = (<DTYPE0_t*>(p + (i+1) * stride))[0]
                    if ai == ai: x[1] += ai
                    ai = (<DTYPE0_t*>(p + (i+2) * stride))[0]
                    if ai == ai: x[2] += ai
                    ai = (<DTYPE0_t*>(p + (i+3) * stride))[0]
                    if ai == ai: x[3] += ai
                asum = (x[0] + x[1]) + (x[2] + x[3])
                for j in range(i + 4, length):
                    ai = (<DTYPE0_t*>(p + j * stride))[0]
                    if ai == ai: asum += ai
                (<DTYPE0_t*>((<char*>pid(ity))))[0] = asum
                PyArray_ITER_NEXT(ita)
                PyArray_ITER_NEXT(ity)
```

If I replace the x[n] with asum then there is no performance gain!

```
In [1]: bn.bench2()
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         0.99       1.06
```

I have no idea what is going on.

**Comment by jennolsen84** [comment]
Benchmarking can be tricky, especially on laptops with all sorts of power management, etc.  I can run the benchmark with my machine.  Do you want me to just run this branch and report?

Also, it might be worth having special code for `stride==1`, perhaps we can get the compiler to auto-vectorize.  

I am not sure what you meant in your comment about x[n].

But, if I understand correctly, what that will do is add dependency for each item in the loop (also the nan comparison is adding a dependency as well) and the code will no longer be able to run in parallel on the processor (see https://en.wikipedia.org/wiki/Superscalar_processor).  I might be 100% correct, but that's probably what is happening.

Also, if you have any special instructions on how to build and run this isolated benchmark, please let me know.

**Comment by kwgoodman** [comment]
Our build system has changed. Once you have the unroll branch you can install inplace with:
```
$ python setup.py build_ext --inplace
```
Or you can use `make all` which will also run the unit tests and flake8. Either way you need cython.

To run the benchmark: `bn.bench2()`.

**Comment by kwgoodman** [comment]
Oh! Interesting that even if I don't set up parallel code the CPU might do it anyway. That would explain things.

Can stride==1? `p` is a pointer to (one-byte) char. So stride==1 would be for a nonexistent one-byte dtype. But if we are moving along the fast axis then I know the stride at compile time if the array is continuous. That's because I have a separate function for each input array dtype. But too early to optimize for axis/continuous etc.

My asum comment was as clear as mud. A diff is worth a thousand words:
```
diff --git a/bottleneck/template/reduce2.pyx b/bottleneck/template/reduce2.pyx
index 084a42f..0eac16e 100644
--- a/bottleneck/template/reduce2.pyx
+++ b/bottleneck/template/reduce2.pyx
@@ -222,7 +222,7 @@ cdef ndarray nansum2_one_DTYPE0(np.flatiter ita,
     # bn.dtypes = [['float64'], ['float32']]
     cdef Py_ssize_t i, j
     cdef DTYPE0_t ai, asum=0
-    cdef DTYPE0_t x[4]
+    # cdef DTYPE0_t x[4]
     cdef ndarray y = PyArray_EMPTY(a_ndim - 1, y_dims, NPY_DTYPE0, 0)
     cdef np.flatiter ity = PyArray_IterNew(y)
     cdef char *p
@@ -233,29 +233,26 @@ cdef ndarray nansum2_one_DTYPE0(np.flatiter ita,
                 PyArray_ITER_NEXT(ity)
         elif length >= 8:
             while PyArray_ITER_NOTDONE(ita):
-                x[0] = 0
-                x[1] = 0
-                x[2] = 0
-                x[3] = 0
+                asum = 0
                 p = <char*>pid(ita)
                 ai = (<DTYPE0_t*>p)[0]
-                if ai == ai: x[0] += ai
+                if ai == ai: asum += ai
                 ai = (<DTYPE0_t*>(p + 1 * stride))[0]
-                if ai == ai: x[1] += ai
+                if ai == ai: asum += ai
                 ai = (<DTYPE0_t*>(p + 2 * stride))[0]
-                if ai == ai: x[2] += ai
+                if ai == ai: asum += ai
                 ai = (<DTYPE0_t*>(p + 3 * stride))[0]
-                if ai == ai: x[3] += ai
+                if ai == ai: asum += ai
                 for i in range(4, length - (length % 4), 4):
                     ai = (<DTYPE0_t*>(p + i * stride))[0]
-                    if ai == ai: x[0] += ai
+                    if ai == ai: asum += ai
                     ai = (<DTYPE0_t*>(p + (i+1) * stride))[0]
-                    if ai == ai: x[1] += ai
+                    if ai == ai: asum += ai
                     ai = (<DTYPE0_t*>(p + (i+2) * stride))[0]
-                    if ai == ai: x[2] += ai
+                    if ai == ai: asum += ai
                     ai = (<DTYPE0_t*>(p + (i+3) * stride))[0]
-                    if ai == ai: x[3] += ai
-                asum = (x[0] + x[1]) + (x[2] + x[3])
+                    if ai == ai: asum += ai
+                # asum = (x[0] + x[1]) + (x[2] + x[3])
                 for j in range(i + 4, length):
                     ai = (<DTYPE0_t*>(p + j * stride))[0]
                     if ai == ai: asum += ai
```

**Comment by jennolsen84** [comment]
Preliminary results (my system was a busy during this time):

```In [2]: bn.bench2()
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         1.31       1.20

In [3]: bn.bench2()
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         1.33       1.14

In [4]: bn.bench2()
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         1.35       1.13
```

**Comment by jennolsen84** [comment]
as for the `stride == 1`, I meant to say that if the array is contiguous.  oops.

**Comment by kwgoodman** [comment]
I ran the benchmarks on my desktop: 64-bit ubuntu 14.04, i7 2600k. What system did you run on?

For axis zero, where the stride is large, the performance boost is not as good:
```
In [3]: bn.bench2(axis=0)
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=0 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         1.17       1.16
```

How does that result fit with the superscalar theory? My guess: The absolute speed up might be the same but now the runtime is longer as we have to use slower cache so the relative performance is worse:

```
In [4]: a=np.random.rand(1000, 10000)
In [5]: timeit bn.nansum(a, 1)
100 loops, best of 3: 7.88 ms per loop
In [6]: timeit bn.nansum2(a, 1)
100 loops, best of 3: 5.92 ms per loop
In [7]: timeit bn.nansum(a, 0)
100 loops, best of 3: 16.3 ms per loop
In [8]: timeit bn.nansum2(a, 0)
100 loops, best of 3: 15.7 ms per loop
```

So according to those results I'd expect the relative performance to be approx cut in half (7.88/16.3). So instead of 1.4 I'd expect 1.2 and got 1.17. (Making up the theory after the experiment is never good.)

**Comment by jennolsen84** [comment]
I ran mine on i7-4790 @ 3.60GHz/8MB Cache. Ubuntu 16.04 (so gcc 5.3)

I think it might be beneficial to try these out on a simpler program and then port the optimizations.  Since little changes can have a big effect.  

I wrote a program (https://gist.github.com/jennolsen84/b49e0a49ca2aca54908f1ad760b38ab5).  And ran it with different compiler options, surprising results.  The output of the program is timing of two different implementations, and also their ratios.  Basically, the simple code can be well optimized with the right compiler flags.  For gcc, `-O2 -ffast-math -ftree-vectorize` did well.  I think the fast-math is needed for reduction.  See http://llvm.org/docs/Vectorizers.html .

A couple of other things: I had to bump up the input size to get meaningful numbers.  Also, the main thing that was moving was the timing of the simpler implementation.

**Comment by kwgoodman** [comment]
I wasn't looking forward to unrolling a bunch of functions. So I made a general unroll function. It is in this commit https://github.com/kwgoodman/bottleneck/commit/e8d24af4f4754a8737c2933c0354d1ca5efdd932. Performance is the same but it should now be easy to apply to other functions.

Studying a simpler code base is a smart idea. When I added `-ffast-math -ftree-vectorize` to nansum2 (but not nansum) I got a nice speed up:
```
In [1]: bn.bench2()
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         2.14       1.81
```

The extra flags gave me a unit test failure (that's odd):
```
func nansum | input a5 (float64) | shape (2,) | axis -1

Input array:
[ inf  nan]

x and y nan location mismatch:
 x: array(nan)
 y: array(inf)
```

I should try some of those flags with the rest of bottleneck.

**Comment by kwgoodman** [comment]
The loop unrolling is a pain to implement and maintain. Better to leave that to the compiler. The key to speed is the independence of the lines of code. So I tried this:

```
            while PyArray_ITER_NOTDONE(ita):
                x[0] = x[1] = x[2] = x[3] = 0
                for i in range(4 * repeat):
                    ai = (<DTYPE0_t*>((<char*>pid(ita)) + i*stride))[0]
                    if ai == ai:
                        x[i % 4] += ai
                asum = x[0] + x[1] + x[2] + x[3]
                for j in range(repeat, left):
                    ai = (<DTYPE0_t*>((<char*>pid(ita)) + j*stride))[0]
                    if ai == ai:
                        asum += ai
                (<DTYPE0_t*>((<char*>pid(ity))))[0] = asum
                PyArray_ITER_NEXT(ita)
                PyArray_ITER_NEXT(ity)
```

The lines in the first for loop are independent for each `i`. Unfortunately performance is not as good as the manual unroll. Using `-funroll-loops` helps but it is still slower. My guess is that `i % 4` is to blame.

Too bad. It would have made things easier.

**Comment by jennolsen84** [comment]
which version of gcc are you using?  Do you have access to docker?  If so, perhaps you can easily try out gcc 5.3, it might make a difference.

The `nansum2` vs `nansum` difference is odd.  As you saw from the test, the best performance for me was the simpler code w/o manual loop unrolling.

**Comment by jennolsen84** [comment]
also when you say:

"When I added -ffast-math -ftree-vectorize to nansum2 (but not nansum) I got a nice speed up", does that mean that you only applied the fast-math and tree-vectorize optimization to nansum2? or you only saw the speedup in nansum2?

**Comment by jennolsen84** [comment]
yeah, after your comment about unit test failure, I would highly recommend trying a newer compiler.  If we find a compiler bug, we'll have to track down which version fixed it, and perhaps disable the optimizations for compiler versions before that.

**Comment by jennolsen84** [comment]
I think it might be good to focus on timeit results instead of ratios, that way if you change compiler flags which effect both functions, you can still do apples-to-apples with older runs.

Or perhaps when we're using ratios, we can just compare against against np.nansum.

**Comment by jennolsen84** [comment]
OK, so I studied your code (should've done that first), and it all makes sense.  I know how you applied the optimization to only reduce2 (saw setup.py), and when I uncommented it out, I still got the failures with gcc 5.3.  

Anyway, I think I know why the manual unroll is out performing the simple case:

the difference is that the pid() call in the unroll function only happens once for 4 elements, vs every time for the manual case.  Pulling it out yields better performance, see diff @ 

https://github.com/jennolsen84/bottleneck/commit/c53c97105a29964dba83dabdc5f75621a30b2aa7

Note, I only changed the floating point function, for demonstration purposes only.

Here are the numbers after the commit.

```
$ make bench
python -c "import bottleneck; bottleneck.bench2()"
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         1.40       1.16
    nanmean        1.40       1.20
```

**Comment by jennolsen84** [comment]
Also, when iterating for `nansum` with axis=-1, we could probably use `order='K'` to iterate over the whole array one time (meaning we only call pid() once).

This is because `external_loop` flag will be set for the iterator since `PyArray_IterAllButAxis` sets it when passing `-1` as the axis.

Also, it seems like the code might convert everything to C arrays.  I might be wrong about this, as I wasn't trying to study the code, and just noticed it.  Perhaps we should try to get rid of that limitation so that if someone with fortran array wants to call `bn.nansum`, it will still be very fast for them

**Comment by kwgoodman** [comment]
https://github.com/jennolsen84/bottleneck/commit/c53c97105a29964dba83dabdc5f75621a30b2aa7 changes nansum_all (which reduces all axes) it does not change nansum_one (which reduces one axis). If you run `bn.bench2(axis=None)` it should test your change.

**Comment by jennolsen84** [comment]
yeah, then the speedup (due to loop unrolling?) only occurs with the additional compiler flags... I will try to see if we can fix the test error (later today)

**Comment by kwgoodman** [comment]
Wait, you see a speed up compared to master with the default compiler flags, right? And with additional flags you get even more speed. That's what I get on my computer.

**Comment by kwgoodman** [comment]
Oh, you're talking about the speed up of https://github.com/jennolsen84/bottleneck/commit/c53c97105a29964dba83dabdc5f75621a30b2aa7. I thought you meant the speed up of reduce2 in general.

**Comment by kwgoodman** [comment]
Yeah, with the compiler flags there is only a small speed up along the fast axis for the unrolling functions.

Additional compiler flags on nansum (-ffast-math, -ftree-vectorize) but not nansum2:
```
In [1]: a=np.random.rand(1000, 10000)
In [2]: timeit bn.nansum(a, 0)
100 loops, best of 3: 14.7 ms per loop
In [3]: timeit bn.nansum2(a, 0)
100 loops, best of 3: 15.5 ms per loop
In [4]: timeit bn.nansum(a, 1)
100 loops, best of 3: 4.64 ms per loop
In [5]: timeit bn.nansum2(a, 1)
100 loops, best of 3: 5.84 ms per loop
In [6]: timeit bn.nansum(a, None)
100 loops, best of 3: 4.49 ms per loop
In [7]: timeit bn.nansum2(a, None)
100 loops, best of 3: 5.84 ms per loop
```

Additional compiler flags on nansum and nansum2 (-ffast-math, -ftree-vectorize):
```
In [1]: a=np.random.rand(1000, 10000)
In [2]: timeit bn.nansum(a, 0)
100 loops, best of 3: 14.7 ms per loop
In [3]: timeit bn.nansum2(a, 0)
100 loops, best of 3: 15 ms per loop
In [4]: timeit bn.nansum(a, 1)
100 loops, best of 3: 4.66 ms per loop
In [5]: timeit bn.nansum2(a, 1)
100 loops, best of 3: 4.27 ms per loop
In [6]: timeit bn.nansum(a, None)
100 loops, best of 3: 4.51 ms per loop
In [7]: timeit bn.nansum2(a, None)
100 loops, best of 3: 4.26 ms per loop
```

My plan was to unroll first and then optimize flags later. You changed the plan: optimize flags first. Then perhaps optimize reduction along the non-fast axis later.

Ugh, I did so much typing to unroll those functions!

**Comment by kwgoodman** [comment]
I segfault in master with:
```
kwargs['reduce']['extra_compile_args'] = ["-ffast-math"]
```

The `-ffast-math` flag ignores (or something like that) NaNs and Inf. Turning back on the NaN and Inf checks with
```
kwargs['reduce']['extra_compile_args'] = ["-ffast-math", "-fno-finite-math-only"]
```
gives:
```
In [1]: a=np.random.rand(1000, 10000)
In [2]: timeit bn.nansum(a, 0)
100 loops, best of 3: 16.2 ms per loop
In [3]: timeit bn.nansum(a, 1)
100 loops, best of 3: 7.89 ms per loop
In [4]: timeit bn.nansum(a, None)
100 loops, best of 3: 7.88 ms per loop
```

Not very fast. Adding `-ftree-vectorize`
```
kwargs['reduce']['extra_compile_args'] = ["-ffast-math", "-fno-finite-math-only", "-ftree-vectorize"]
```
gives worse performance.

I think part of the speed we initially saw with "-ffast-math", "-ftree-vectorize" is due to ignoring NaNs and Inf.

Maybe we could use "-ffast-math", "-ftree-vectorize" when the input array contain ints (and therefore no nans, inf). I don't know if we can turn that on and off in cython code. If not maybe we could split out a reduce_int.pyx or something.

**Comment by jennolsen84** [comment]
sometimes with optimization, I find you end up doing things multiple ways to find out what's best.

Here is where I am at now (all test) fath:


```
In [2]: bn.bench2(axis=None)
Bottleneck unroll performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is rolled time divided by unrolled time
    NaN means approx one-third NaNs; float64 and axis=None are used

                 no NaN      NaN    
                (1000,1000)(1000,1000)
    nansum         1.67       1.27
    nanmean        1.53       1.11


```

**Comment by jennolsen84** [comment]
I am also using
```
    kwargs['reduce2']['extra_compile_args'] = ["-ffast-math","-fno-finite-math-only",
                                               "-ftree-vectorize"]
```
but all tests pass for me

**Comment by kwgoodman** [comment]
Yes, all tests pass for me too. But I'm on master, not the reduce2 branch. I thought you were saying that we could get the performance with just the flags instead of manual unrolling (and making statement independent).

**Comment by jennolsen84** [comment]
yes, it seems like we have achieved the performance increase without manual unrolling, or am I missing something?

**Comment by kwgoodman** [comment]
The latest benchmark you posted says bn.bench2. The 2 means you are on the unroll branch.

bn.bench2 is master branch runtime divided by manual unroll time. Larger than 1 means manual unroll helps. And your kwargs['reduce2'] applies only to the manual unroll functions.

**Comment by jennolsen84** [comment]
yeah, I am on my own branch with https://github.com/jennolsen84/bottleneck/commit/c53c97105a29964dba83dabdc5f75621a30b2aa7 applied.  So, we should be OK?

**Comment by kwgoodman** [comment]
Ah! I see. That's a confusing branch.

**Comment by jennolsen84** [comment]
I could make a PR if you'd like

**Comment by kwgoodman** [comment]
Sure, I'd like to test it. How about making it off of master? So changing flags and adding  `cdef char *p` to reduce.pyx. And then using the benchmark in master which use numpy as the baseline.

**Comment by kwgoodman** [comment]
Oh, that wouldn't give us a comparison to unrolled. OK, whatever PR you make I'll take a look another day. I need a break.

**Comment by kwgoodman** [comment]
I don't see a speed up from the `p = <char*>pid(ita)` change. Seems like the flags alone are responsible for the speed up. Do other compilers have those three flags?

We don't need `-fno-finite-math-only` for int input arrays. Not sure if we should also take care of that at the same time.

Anyway, I'll open a new issue: #122.
